src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,784 @@
+            vector<WeightMap>&                  pixel_samples,
+            map<pair<int, int>, NameMap>&       tile_name_map,
+            int                                 num_layers,
+            CryptomatteAOV::CryptomatteType     layer_type)
+            : m_aov_image(aov_image)
+            , m_pixel_samples(pixel_samples)
+            , m_num_layers(num_layers)
+            , m_tile_name_map(tile_name_map)
+            , m_layer_type(layer_type)
+        {
+        }
+
+        void on_tile_begin(
+            const Frame&                frame,
+            const size_t                tile_x,
+            const size_t                tile_y,
+            const size_t                max_spp) override
+        {
+            // Fetch the destination tile.
+            const CanvasProperties& props = frame.image().properties();
+            const Tile& tile = frame.image().tile(tile_x, tile_y);
+
+            m_frame_width = props.m_canvas_width;
+            m_rcp_canvas_width = props.m_rcp_canvas_width;
+            m_rcp_canvas_height = props.m_rcp_canvas_height;
+            m_renderer_filter = &(frame.get_filter());
+
+            // Fetch the tile bounds (inclusive).
+            m_tile_origin_x = static_cast<int>(tile_x * props.m_tile_width);
+            m_tile_origin_y = static_cast<int>(tile_y * props.m_tile_height);
+            m_tile_end_x = static_cast<int>(m_tile_origin_x + tile.get_width() - 1);
+            m_tile_end_y = static_cast<int>(m_tile_origin_y + tile.get_height() - 1);
+
+            for (int ry = m_tile_origin_y; ry <= m_tile_end_y; ++ry)
+            {
+                for (int rx = m_tile_origin_x; rx <= m_tile_end_x; ++rx)
+                {
+                    m_pixel_samples[ry * m_frame_width + rx] = WeightMap();
+                }
+            }
+
+            if (frame.has_crop_window())
+                m_crop_window = frame.get_crop_window();
+            else
+                m_crop_window = AABB2u(Vector2u(m_tile_origin_x, m_tile_origin_y), Vector2u(m_tile_end_x, m_tile_end_y));
+        }
+
+        void on_tile_end(
+            const Frame&                frame,
+            const size_t                tile_x,
+            const size_t                tile_y) override
+        {
+            for (int ry = m_tile_origin_y; ry <= m_tile_end_y; ++ry)
+            {
+                for (int rx = m_tile_origin_x; rx <= m_tile_end_x; ++rx)
+                {
+                    const WeightMap& weight_map = m_pixel_samples[ry * m_frame_width + rx];
+
+                    if (weight_map.size() > 0)
+                    {
+                        float total_weight = 0.0f;
+                        for (const auto& item : weight_map)
+                            total_weight += item.second;
+
+                        if (total_weight == 0.0f)
+                            total_weight = 1.0f;
+
+                        vector<pair<float, uint32_t>> ranked_vector;
```

Comment:
Reserve memory in advance since you know exactly how many items the vector will hold.

Even better would be to have a single vector allocated outside of the loop, and cleared (using `foundation::clear_keep_memory()`) every time. You wouldn't need to reserve memory anymore and you would save a lot of memory allocations and deallocations.